metric_learn/constraints.py

```diff
@@ -33,6 +35,110 @@ def positive_negative_pairs(self, num_constraints, same_length=False,
       return a[:n], b[:n], c[:n], d[:n]
     return a, b, c, d
 
+  def generate_knntriplets(self, X, k_genuine, k_impostor):
+    """
+    Generates triplets from labeled data.
+
+    For every point (X_a) the triplets (X_a, X_b, X_c) are constructed from all
+    the combinations of taking one of its `k_genuine`-nearest neighbors of the
+    same class (X_b) and taking one of its `k_impostor`-nearest neighbors of
+    other classes (X_c).
+
+    In the case a class doesn't have enough points in the same class (other
+    classes) to yield `k_genuine` (`k_impostor`) neighbors a warning will be
+    raised and the maximum value of genuine (impostor) neighbors will be used
+    for that class.
+
+    Parameters
+    ----------
+      X : (n x d) matrix
+        Input data, where each row corresponds to a single instance.
+      k_genuine : int
+        Number of neighbors of the same class to be taken into account.
+      k_impostor : int
+        Number of neighbors of different classes to be taken into account.
+
+    Returns
+    -------
+    triplets : array-like, shape=(n_constraints, 3)
+      2D array of triplets of indicators.
+    """
+
+    labels, labels_count = np.unique(self.partial_labels, return_counts=True)
+    n_labels = len(labels)
+    len_input = np.size(self.partial_labels, 0)
+
+    # Handle the case where there are too few elements to yield k_genuine or
+    # k_impostor neighbors for every class.
+
+    k_genuine_vec = np.ones(n_labels, dtype=np.intp)*k_genuine
+    k_impostor_vec = np.ones(n_labels, dtype=np.intp)*k_impostor
+
+    for i in range(n_labels):
+      if (k_genuine + 1 > labels_count[i]):
+        k_genuine_vec[i] = labels_count[i]-1
+        warnings.warn("The class {} has {} elements, which is not sufficient "
+                      "to generate {} genuine neighbors as specified by "
+                      "k_genuine. Will generate {} genuine neighbors instead."
+                      "\n"
+                      .format(labels[i], labels_count[i], k_genuine+1,
+                              k_genuine_vec[i]))
+      if (k_impostor > len_input - labels_count[i]):
+        k_impostor_vec[i] = len_input - labels_count[i]
+        warnings.warn("The class {} has {} elements of other classes, which is"
+                      " not sufficient to generate {} impostor neighbors as "
+                      "specified by k_impostor. Will generate {} impostor "
+                      "neighbors instead.\n"
+                      .format(labels[i], k_impostor_vec[i], k_impostor,
+                              k_impostor_vec[i]))
+
+    triplets = np.empty((np.dot(k_genuine_vec*k_impostor_vec, labels_count),
```

Comment:
This is confusing. Let's split it up:

```
num_triplets = labels_count.dot(k_genuine_vec * k_impostor_vec)
triplets = np.empty((num_triplets, 3), dtype=np.intp)
```

Also a comment explaining the math would be useful.